File: packages/Auto-claude-code-research-in-sleep/tools/experiment_env/env_helper.py
#!/usr/bin/env python3
"""env_helper.py — CLI entrypoint for the experiment_env helper.

Subcommands:
    parse     validate agent's candidate JSON + write .aris/experiment-env.json
    info      print the current env config (env_type, fields)
    provision create/rent the environment
    preflight GPU/conda/connectivity check
    sync      sync code to the environment
    deploy    launch one job (--run-spec PATH)
    monitor   query job status (--handle PATH)
    collect   download results (--handle PATH)
    destroy   tear the environment down (--handle PATH)

Every action subcommand: reads --env-config (default
.aris/experiment-env.json), checks staleness (source mtime + sha256),
constructs EnvBackend.create(env_type, config), dispatches, prints JSON
to stdout. exit 0 ok / 1 hard error / 2 no config / 10-16 method error.

Layer 0-3 resolution block (copy into each consuming SKILL.md):

    ENV_HELPER=""
    if [ -z "${ARIS_REPO:-}" ] && [ -f .aris/installed-skills.txt ]; then
        ARIS_REPO=$(awk -F'\t' '$1=="repo_root"{print $2; exit}' .aris/installed-skills.txt 2>/dev/null) || true
    fi
    ENV_HELPER=".aris/tools/experiment_env/env_helper.py"
    [ -f "$ENV_HELPER" ] || ENV_HELPER="tools/experiment_env/env_helper.py"
    [ -f "$ENV_HELPER" ] || { [ -n "${ARIS_REPO:-}" ] && ENV_HELPER="$ARIS_REPO/tools/experiment_env/env_helper.py"; }
    [ -f "$ENV_HELPER" ] || ENV_HELPER=""
    [ -z "$ENV_HELPER" ] && { echo "ERROR: experiment_env helper not found" >&2; exit 1; }
"""

import argparse
import hashlib
import json
import sys
from pathlib import Path

# Make `import parse_env` / `from env_backend import ...` work whether
# invoked as `python3 tools/experiment_env/env_helper.py` (script) or
# `python3 -m tools.experiment_env.env_helper` (module).
HERE = Path(__file__).resolve().parent
if str(HERE) not in sys.path:
    sys.path.insert(0, str(HERE))

try:  # package mode (python -m tools.experiment_env.env_helper)
    from . import parse_env  # noqa: E402
    from .env_backend import EnvBackend, EnvError  # noqa: E402
except ImportError:  # script mode (python3 tools/experiment_env/env_helper.py)
    import parse_env  # noqa: E402
    from env_backend import EnvBackend, EnvError  # noqa: E402

DEFAULT_CONFIG = ".aris/experiment-env.json"


def _load_config(env_config: str) -> dict:
    p = Path(env_config)
    if not p.exists():
        sys.stderr.write(
            f"ERROR: env config not found at {env_config}\n"
            f"       Run `env_helper.py parse --json <candidate> --source "
            f"CLAUDE.md` first (the agent produces the candidate JSON from "
            f"CLAUDE.md/AGENTS.md).\n"
        )
        sys.exit(2)
    return json.loads(p.read_text())


def _check_stale(cfg: dict, env_config: str):
    """Warn (non-blocking) if the source markdown changed since last parse."""
    source = cfg.get("source_path")
    if not source or not Path(source).exists():
        return
    sp = Path(source)
    try:
        st = sp.stat()
        content = sp.read_bytes()
    except OSError:
        return
    mtime = int(st.st_mtime)
    digest = hashlib.sha256(content).hexdigest()
    stale = (cfg.get("source_mtime") != mtime) or (cfg.get("source_hash") != digest)
    if stale:
        sys.stderr.write(
            f"WARN: {source} changed since last parse (mtime/hash mismatch); "
            f"re-run `env_helper.py parse` to refresh {env_config}\n"
        )


def _backend_from_config(env_config: str, dry_run: bool) -> "EnvBackend":
    cfg = _load_config(env_config)
    _check_stale(cfg, env_config)
    env_type = cfg.get("env_type")
    if not env_type:
        sys.stderr.write("ERROR: env_config has no env_type\n")
        sys.exit(1)
    block = cfg.get(env_type)
    if not block:
        sys.stderr.write(f"ERROR: env_type={env_type!r} but no {env_type!r} "
                         f"block in config\n")
        sys.exit(1)
    state_dir = str(Path(env_config).resolve().parent)
    return EnvBackend.create(env_type, block, state_dir=state_dir, dry_run=dry_run)


def _emit(result):
    json.dump(result, sys.stdout, indent=2, ensure_ascii=False)
    sys.stdout.write("\n")


def _read_json_file(path: str) -> dict:
    return json.loads(Path(path).read_text())


def cmd_parse(args):
    candidate = _read_json_file_stdin(args.json)
    try:
        validated = parse_env.validate(
            candidate, source_path=args.source,
            env_type_override=args.env_type)
        if not args.stdout_only:
            parse_env.write_config(validated, args.out)
        _emit(validated)
    except parse_env.ValidationError as e:
        sys.stderr.write(f"ERROR: {e.message}\n")
        return 1
    return 0


def _read_json_file_stdin(spec):
    if spec == "-":
        return json.loads(sys.stdin.read())
    return json.loads(Path(spec).read_text())


def cmd_info(args):
    cfg = _load_config(args.env_config)
    env_type = cfg.get("env_type")
    block = cfg.get(env_type, {}) if env_type else {}
    _emit({"env_type": env_type, "source": cfg.get("source"),
           "parsed_at": cfg.get("parsed_at"),
           "warnings": cfg.get("warnings", []),
           "fields": block})
    return 0


def cmd_provision(args):
    b = _backend_from_config(args.env_config, args.dry_run)
    return _run_method(b.provision, 10)


def cmd_preflight(args):
    b = _backend_from_config(args.env_config, args.dry_run)
    return _run_method(b.preflight, 11)


def cmd_sync(args):
    b = _backend_from_config(args.env_config, args.dry_run)
    return _run_method(lambda: b.sync(args.src), 12)


def cmd_deploy(args):
    b = _backend_from_config(args.env_config, args.dry_run)
    run_spec = _read_json_file(args.run_spec)
    return _run_method(lambda: b.deploy(run_spec), 13)


def cmd_monitor(args):
    b = _backend_from_config(args.env_config, args.dry_run)
    handle = _read_json_file(args.handle)
    return _run_method(lambda: b.monitor(handle), 14)


def cmd_collect(args):
    b = _backend_from_config(args.env_config, args.dry_run)
    return _run_method(b.collect_results, 15)


def cmd_destroy(args):
    b = _backend_from_config(args.env_config, args.dry_run)
    return _run_method(b.destroy, 16)


def _run_method(fn, err_code):
    try:
        result = fn()
        _emit(result)
        return 0
    except EnvError as e:
        sys.stderr.write(f"ERROR: {e.message}\n")
        return e.code
    except KeyError as e:
        sys.stderr.write(f"ERROR: missing field {e}\n")
        return err_code


def build_parser():
    p = argparse.ArgumentParser(
        prog="env_helper.py",
        description="Unified experiment-environment helper (ARIS).",
    )
    sub = p.add_subparsers(dest="cmd", required=True)

    pp = sub.add_parser("parse", help="validate candidate JSON + write config")
    pp.add_argument("--json", required=True, help="candidate JSON path or '-' for stdin")
    pp.add_argument("--source", default=None, help="CLAUDE.md/AGENTS.md path")
    pp.add_argument("--env-type", default=None)
    pp.add_argument("--out", default=DEFAULT_CONFIG)
    pp.add_argument("--stdout-only", action="store_true")
    pp.set_defaults(func=cmd_parse)

    pi = sub.add_parser("info", help="print current env config")
    pi.add_argument("--env-config", default=DEFAULT_CONFIG)
    pi.set_defaults(func=cmd_info)

    for name, fn, extra in [
        ("provision", cmd_provision, []),
        ("preflight", cmd_preflight, []),
        ("sync", cmd_sync, [("--src", {"required": True, "help": "local source dir"})]),
        ("deploy", cmd_deploy, [("--run-spec", {"required": True, "help": "path to run_spec JSON"})]),
        ("monitor", cmd_monitor, [("--handle", {"required": True, "help": "path to handle JSON"})]),
        ("collect", cmd_collect, []),
        ("destroy", cmd_destroy, [("--handle", {"required": False, "help": "path to handle JSON"})]),
    ]:
        sp = sub.add_parser(name, help=f"{name} the environment")
        sp.add_argument("--env-config", default=DEFAULT_CONFIG)
        for flag, kw in extra:
            sp.add_argument(flag, **kw)
        sp.add_argument("--dry-run", action="store_true",
                        help="print what would run, do not execute")
        sp.set_defaults(func=fn)
    return p


def main(argv=None):
    p = build_parser()
    args = p.parse_args(argv)
    return args.func(args)


if __name__ == "__main__":
    sys.exit(main())

File: packages/Auto-claude-code-research-in-sleep/src/tools/experiment-env/env-helper.ts
#!/usr/bin/env node
import crypto from "crypto";
import fs from "fs";
import path from "path";
import { createCli, runCli } from "../../lib/cli.js";
import { EnvBackend, EnvError } from "./env-backend.js";
import { validate, writeConfig, ValidationError } from "./parse-env.js";

const DEFAULT_CONFIG = ".aris/experiment-env.json";

function loadConfig(envConfig: string): Record<string, unknown> {
  if (!fs.existsSync(envConfig)) {
    process.stderr.write(
      `ERROR: env config not found at ${envConfig}\n` +
        `       Run \`env_helper parse --json <candidate> --source CLAUDE.md\` first.\n`,
    );
    process.exit(2);
  }
  return JSON.parse(fs.readFileSync(envConfig, "utf-8")) as Record<string, unknown>;
}

function checkStale(cfg: Record<string, unknown>, envConfig: string): void {
  const source = cfg.source_path as string | undefined;
  if (!source || !fs.existsSync(source)) return;
  try {
    const st = fs.statSync(source);
    const content = fs.readFileSync(source);
    const mtime = Math.floor(st.mtimeMs / 1000);
    const digest = crypto.createHash("sha256").update(content).digest("hex");
    const stale = cfg.source_mtime !== mtime || cfg.source_hash !== digest;
    if (stale) {
      process.stderr.write(
        `WARN: ${source} changed since last parse (mtime/hash mismatch); ` +
          `re-run \`env_helper parse\` to refresh ${envConfig}\n`,
      );
    }
  } catch {
    // ignore read errors
  }
}

function backendFromConfig(envConfig: string, dryRun: boolean): EnvBackend {
  const cfg = loadConfig(envConfig);
  checkStale(cfg, envConfig);
  const envType = cfg.env_type as string | undefined;
  if (!envType) {
    process.stderr.write("ERROR: env_config has no env_type\n");
    process.exit(1);
  }
  const block = cfg[envType] as Record<string, unknown> | undefined;
  if (!block) {
    process.stderr.write(`ERROR: env_type='${envType}' but no '${envType}' block in config\n`);
    process.exit(1);
  }
  const stateDir = path.dirname(path.resolve(envConfig));
  return EnvBackend.create(envType, block, stateDir, dryRun);
}

function emit(result: unknown): void {
  console.log(JSON.stringify(result, null, 2));
}

function readJsonFileStdin(spec: string): Record<string, unknown> {
  if (spec === "-") {
    return JSON.parse(fs.readFileSync(0, "utf-8")) as Record<string, unknown>;
  }
  return JSON.parse(fs.readFileSync(spec, "utf-8")) as Record<string, unknown>;
}

function runMethod(fn: () => Record<string, unknown>, errCode: number): void {
  try {
    const result = fn();
    emit(result);
  } catch (e) {
    if (e instanceof EnvError) {
      process.stderr.write(`ERROR: ${e.message}\n`);
      process.exit(e.code);
    }
    if (e instanceof Error && e.message.includes("missing")) {
      process.stderr.write(`ERROR: ${e.message}\n`);
      process.exit(errCode);
    }
    throw e;
  }
}

const program = createCli("env-helper", "Unified experiment-environment helper (ARIS)");

program
  .command("parse")
  .description("validate candidate JSON + write config")
  .requiredOption("--json <path>", "candidate JSON path or '-' for stdin")
  .option("--source <path>", "CLAUDE.md/AGENTS.md path")
  .option("--env-type <type>", "override env_type in the candidate")
  .option("--out <path>", "output path", DEFAULT_CONFIG)
  .option("--stdout-only", "do not write file, just print validated JSON")
  .action(
    (opts: {
      json: string;
      source?: string;
      envType?: string;
      out: string;
      stdoutOnly?: boolean;
    }) => {
      try {
        const candidate = readJsonFileStdin(opts.json);
        const validated = validate(candidate, opts.source, opts.envType);
        if (!opts.stdoutOnly) {
          writeConfig(validated, opts.out);
        }
        emit(validated);
      } catch (e) {
        if (e instanceof ValidationError) {
          process.stderr.write(`ERROR: ${e.message}\n`);
          process.exit(1);
        }
        throw e;
      }
    },
  );

program
  .command("info")
  .description("print current env config")
  .option("--env-config <path>", "env config path", DEFAULT_CONFIG)
  .action((opts: { envConfig: string }) => {
    const cfg = loadConfig(opts.envConfig);
    const envType = cfg.env_type as string | undefined;
    const block = envType ? (cfg[envType] as Record<string, unknown>) || {} : {};
    emit({
      env_type: envType,
      source: cfg.source,
      parsed_at: cfg.parsed_at,
      warnings: cfg.warnings || [],
      fields: block,
    });
  });

for (const { name, extra, errCode } of [
  { name: "provision", extra: [] as string[], errCode: 10 },
  { name: "preflight", extra: [] as string[], errCode: 11 },
  { name: "sync", extra: ["src"] as string[], errCode: 12 },
  { name: "deploy", extra: ["run-spec"] as string[], errCode: 13 },
  { name: "monitor", extra: ["handle"] as string[], errCode: 14 },
  { name: "collect", extra: [] as string[], errCode: 15 },
  { name: "destroy", extra: ["handle-optional"] as string[], errCode: 16 },
]) {
  const cmd = program
    .command(name)
    .description(`${name} the environment`)
    .option("--env-config <path>", "env config path", DEFAULT_CONFIG)
    .option("--dry-run", "print what would run, do not execute");

  if (extra.includes("src")) {
    cmd.requiredOption("--src <path>", "local source dir");
  }
  if (extra.includes("run-spec")) {
    cmd.requiredOption("--run-spec <path>", "path to run_spec JSON");
  }
  if (extra.includes("handle")) {
    cmd.requiredOption("--handle <path>", "path to handle JSON");
  }
  if (extra.includes("handle-optional")) {
    cmd.option("--handle <path>", "path to handle JSON");
  }

  cmd.action(
    (opts: {
      envConfig: string;
      dryRun?: boolean;
      src?: string;
      runSpec?: string;
      handle?: string;
    }) => {
      const b = backendFromConfig(opts.envConfig, !!opts.dryRun);
      switch (name) {
        case "provision":
          runMethod(() => b.provision(), errCode);
          break;
        case "preflight":
          runMethod(() => b.preflight(), errCode);
          break;
        case "sync":
          runMethod(() => b.sync(opts.src!), errCode);
          break;
        case "deploy": {
          const runSpec = JSON.parse(fs.readFileSync(opts.runSpec!, "utf-8")) as Record<
            string,
            unknown
          >;
          runMethod(() => b.deploy(runSpec), errCode);
          break;
        }
        case "monitor": {
          const handle = JSON.parse(fs.readFileSync(opts.handle!, "utf-8")) as Record<
            string,
            unknown
          >;
          runMethod(() => b.monitor(handle), errCode);
          break;
        }
        case "collect":
          runMethod(() => b.collectResults(), errCode);
          break;
        case "destroy":
          runMethod(() => b.destroy(), errCode);
          break;
      }
    },
  );
}

runCli(program);
